src/NuGet.Core/NuGet.Packaging/PackageFileExtractor.cs

```diff
@@ -0,0 +1,122 @@
+                f.StartsWith("ref/", StringComparison.OrdinalIgnoreCase))
+                .ToArray();
+
+            foreach (var file in refAndLibFiles)
+            {
+                if (!file.EndsWith(".xml"))
+                {
+                    continue;
+                }
+
+                // Xml files located next to the neutral language dll.
+                var dllFile = Path.ChangeExtension(file, ".dll");
+                if (refAndLibFiles.Contains(dllFile, StringComparer.OrdinalIgnoreCase))
+                {
+                    intellisenseXmlFiles.Add(file);
+                    continue;
+                }
+
+                // For satellite assemblies, look for a corresponding .resources.dll
+                var resourceDll = Path.ChangeExtension(file, ".resources.dll");
+                if (refAndLibFiles.Contains(resourceDll, StringComparer.OrdinalIgnoreCase))
+                {
+                    intellisenseXmlFiles.Add(file);
+                    continue;
+                }
+
+                // Xml files located in a language specific directory.
+                var languageDllFile = GetBinaryForLanguageSpecificXml(file);
+                if (languageDllFile != null && refAndLibFiles.Contains(languageDllFile, StringComparer.Ordinal))
+                {
+                    intellisenseXmlFiles.Add(file);
+                }
+            }
+
+            return intellisenseXmlFiles;
+        }
+
+        private static string GetBinaryForLanguageSpecificXml(string file)
+        {
+            // For xml files located in language specific directories, look for a corresponding binary
+            // in the parent directory.
+            // e.g. ref/net45/fr/MyBinary.xml -> ref/net45/MyBinary.dll
+
+            var fileSeparatorIndex = file.LastIndexOf('/');
+            if (fileSeparatorIndex > 0)
+            {
+                var directorySeparatorIndex = file.LastIndexOf('/', fileSeparatorIndex - 1);
+                if (directorySeparatorIndex >= 0)
+                {
+                    var fileAtParentDirectory =
+                        file.Substring(0, directorySeparatorIndex) +
+                        file.Substring(fileSeparatorIndex);
+                    return Path.ChangeExtension(fileAtParentDirectory, ".dll");
+                }
+            }
+
+            return null;
+        }
+
+        public string ExtractPackageFile(string source, string target, Stream stream)
+        {
+            if ((_xmlDocFileSaveMode == XmlDocFileSaveMode.Skip) && _intellisenseXmlFiles.Contains(source))
+            {
+                return null;
+            }
+
+            var extractDirectory = Path.GetDirectoryName(target);
+            Directory.CreateDirectory(extractDirectory);
+
+            if ((_xmlDocFileSaveMode == XmlDocFileSaveMode.Compress) && _intellisenseXmlFiles.Contains(source))
+            {
+                var targetZip = Path.ChangeExtension(target, ".xml.zip");
```

Comment:
The package could potentially include both the .xml and .xml.zip file already which looks like it could blow up here. Either one could end up on disk first so this should handle both scenarios.